File: packages/sdk-go/freeze_test.go
package rateguard

import (
	"io"
	"net/http"
	"net/http/httptest"
	"strings"
	"testing"
)

func freezeSend(t *testing.T, client *http.Client, customer string) *http.Response {
	t.Helper()
	req, _ := http.NewRequest(http.MethodPost, "https://api.openai.com/v1/chat/completions",
		strings.NewReader(`{"model":"gpt-4o"}`))
	if customer != "" {
		req.Header.Set("X-RateGuard-Customer", customer)
	}
	resp, err := client.Do(req)
	if err != nil {
		t.Fatalf("outbound call failed: %v", err)
	}
	_, _ = io.ReadAll(resp.Body)
	_ = resp.Body.Close()
	return resp
}

func TestFreezeGlobalHaltsThenResumes(t *testing.T) {
	server := httptest.NewServer(openAIJSONHandler("gpt-4o", 100, 50))
	defer server.Close()
	sdk := New(Config{Preset: "dev", TokenBudgetPerHour: 100_000})
	client := wrapForHost(t, sdk, server)

	if resp := freezeSend(t, client, ""); resp.StatusCode != http.StatusOK {
		t.Fatalf("before freeze: want 200, got %d", resp.StatusCode)
	}

	sdk.Freeze("")
	if !sdk.IsFrozen("") {
		t.Error("IsFrozen(\"\") should be true after a global freeze")
	}
	resp := freezeSend(t, client, "")
	if resp.StatusCode != http.StatusForbidden {
		t.Fatalf("frozen: want 403, got %d", resp.StatusCode)
	}
	if resp.Header.Get("X-RateGuard-Synthesized") != "true" {
		t.Error("frozen response should be marked synthesized")
	}

	sdk.Unfreeze("")
	if resp := freezeSend(t, client, ""); resp.StatusCode != http.StatusOK {
		t.Fatalf("after unfreeze: want 200, got %d", resp.StatusCode)
	}
}

func TestFreezePerCustomerIsScoped(t *testing.T) {
	server := httptest.NewServer(openAIJSONHandler("gpt-4o", 100, 50))
	defer server.Close()
	sdk := New(Config{Preset: "dev", TokenBudgetPerHour: 100_000})
	client := wrapForHost(t, sdk, server)

	sdk.Freeze("alice")
	if got := freezeSend(t, client, "alice").StatusCode; got != http.StatusForbidden {
		t.Fatalf("alice frozen: want 403, got %d", got)
	}
	if got := freezeSend(t, client, "bob").StatusCode; got != http.StatusOK {
		t.Fatalf("bob is a different customer: want 200, got %d", got)
	}
	if scopes := sdk.FrozenScopes(); len(scopes) != 1 || scopes[0] != "customer=alice" {
		t.Errorf("FrozenScopes = %v, want [customer=alice]", scopes)
	}
}

func TestFreezeIgnoredInObserveMode(t *testing.T) {
	server := httptest.NewServer(openAIJSONHandler("gpt-4o", 100, 50))
	defer server.Close()
	sdk := New(Config{Preset: "dev", TokenBudgetPerHour: 100_000})
	client := wrapForHost(t, sdk, server, OutboundOptions{Mode: OutboundModeObserve})

	sdk.Freeze("")
	if got := freezeSend(t, client, "").StatusCode; got != http.StatusOK {
		t.Errorf("observe mode never blocks, even frozen: got %d", got)
	}
}

func TestAdminFreezeEndpoint(t *testing.T) {
	sdk := New(Config{Preset: "dev"})
	admin := sdk.AdminHandler()

	do := func(method, path, body string) *httptest.ResponseRecorder {
		rr := httptest.NewRecorder()
		admin.ServeHTTP(rr, httptest.NewRequest(method, path, strings.NewReader(body)))
		return rr
	}

	if rr := do(http.MethodPost, "/admin/freeze", `{"scope":"alice"}`); rr.Code != http.StatusOK {
		t.Fatalf("POST /admin/freeze: want 200, got %d", rr.Code)
	}
	if !sdk.IsFrozen("alice") {
		t.Error("alice should be frozen after the admin POST")
	}
	if rr := do(http.MethodGet, "/admin/frozen", ""); !strings.Contains(rr.Body.String(), "customer=alice") {
		t.Errorf("/admin/frozen should list alice, got %s", rr.Body.String())
	}
	if rr := do(http.MethodPost, "/admin/unfreeze", `{"scope":"alice"}`); rr.Code != http.StatusOK {
		t.Fatalf("POST /admin/unfreeze: want 200, got %d", rr.Code)
	}
	if sdk.IsFrozen("alice") {
		t.Error("alice should be unfrozen after the admin POST")
	}
}

File: packages/sdk-node/test/freeze.test.ts
import { describe, it, expect } from 'vitest';
import { RateGuard } from '../src/index.js';

const okResponse = () =>
  new Response(
    JSON.stringify({ model: 'gpt-4o', usage: { prompt_tokens: 100, completion_tokens: 50, total_tokens: 150 } }),
    { status: 200, headers: { 'content-type': 'application/json' } },
  );

function send(wrapped: typeof fetch, customer?: string) {
  return wrapped('https://api.openai.com/v1/chat/completions', {
    method: 'POST',
    body: JSON.stringify({ model: 'gpt-4o' }),
    ...(customer ? { headers: { 'X-RateGuard-Customer': customer } } : {}),
  });
}

describe('freeze kill switch', () => {
  it('global freeze halts, unfreeze resumes', async () => {
    const rg = new RateGuard({ preset: 'dev', tokenBudget: { hourLimit: 100_000 } });
    const wrapped = rg.wrapFetch({ fetch: (async () => okResponse()) as typeof fetch });

    expect((await send(wrapped)).status).toBe(200);

    rg.freeze();
    expect(rg.isFrozen()).toBe(true);
    const blocked = await send(wrapped);
    expect(blocked.status).toBe(403);
    expect(blocked.headers.get('x-rateguard-synthesized')).toBe('true');

    rg.unfreeze();
    expect((await send(wrapped)).status).toBe(200);
  });

  it('per-customer freeze is scoped to that customer', async () => {
    const rg = new RateGuard({ preset: 'dev', tokenBudget: { hourLimit: 100_000 } });
    const wrapped = rg.wrapFetch({ fetch: (async () => okResponse()) as typeof fetch });

    rg.freeze('alice');
    expect((await send(wrapped, 'alice')).status).toBe(403);
    expect((await send(wrapped, 'bob')).status).toBe(200);
    expect(rg.frozenScopes()).toEqual(['customer=alice']);
  });

  it('observe mode never blocks, even when frozen', async () => {
    const rg = new RateGuard({ preset: 'dev', tokenBudget: { hourLimit: 100_000 } });
    const wrapped = rg.wrapFetch({ mode: 'observe', fetch: (async () => okResponse()) as typeof fetch });

    rg.freeze();
    expect((await send(wrapped)).status).toBe(200);
  });
});
